rocketpy/motors/Tank.py

```diff
@@ -0,0 +1,626 @@
+    def gasHeight(self):
+        fluid_volume = self.gasVolume + self.liquidVolume
+        gasHeight = self.geometry.inverse_volume.compose(fluid_volume)
+        if (gasHeight > self.geometry.top).any():
+            raise ValueError(f"The tank {self.name} is overfilled.")
+        return gasHeight
+
+    def discretize_flow(self):
+        self.liquid_mass_flow_rate_in.setDiscrete(*self.flux_time, self.discretize)
+        self.gas_mass_flow_rate_in.setDiscrete(*self.flux_time, self.discretize)
+        self.liquid_mass_flow_rate_out.setDiscrete(*self.flux_time, self.discretize)
+        self.gas_mass_flow_rate_out.setDiscrete(*self.flux_time, self.discretize)
+
+
+class UllageBasedTank(Tank):
+    def __init__(
+        self,
+        name,
+        geometry,
+        flux_time,
+        liquid,
+        gas,
+        ullage,
+        discretize=100,
+    ):
+        super().__init__(name, geometry, flux_time, gas, liquid, discretize)
+
+        # Define ullage
+        self.ullage = Function(ullage, "Time (s)", "Volume (m³)", "linear")
+
+        # Discretize input if needed
+        self.discretize_ullage() if discretize else None
+
+        # Check if the ullage is within bounds
+        if (self.ullage > self.geometry.total_volume).any() or (self.ullage < 0).any():
+            raise ValueError("The ullage volume is out of bounds.")
+
+    @funcify_method("Time (s)", "Mass (kg)")
+    def mass(self):
+        return self.liquidMass + self.gasMass
+
+    @funcify_method("Time (s)", "mass flow rate (kg/s)")
+    def netMassFlowRate(self):
+        return self.mass.derivativeFunction()
+
+    @funcify_method("Time (s)", "Volume (m³)")
+    def liquidVolume(self):
+        return -(self.ullage - self.geometry.total_volume)
+
+    @funcify_method("Time (s)", "Volume (m³)")
+    def gasVolume(self):
+        return self.ullage
+
+    @funcify_method("Time (s)", "Mass (kg)")
+    def gasMass(self):
+        return self.gasVolume * self.gas.density
+
+    @funcify_method("Time (s)", "Mass (kg)")
+    def liquidMass(self):
+        return self.liquidVolume * self.liquid.density
+
+    @funcify_method("Time (s)", "Height (m)")
+    def liquidHeight(self):
+        return self.geometry.inverse_volume.compose(self.liquidVolume)
+
+    @funcify_method("Time (s)", "Height (m)", "linear")
+    def gasHeight(self):
+        return Function(self.geometry.top).setDiscreteBasedOnModel(self.gasVolume)
+
+    def discretize_ullage(self):
+        self.ullage.setDiscrete(*self.flux_time, self.discretize)
+
+
+class LevelBasedTank(Tank):
+    def __init__(
+        self,
+        name,
+        geometry,
+        flux_time,
+        liquid,
+        gas,
+        liquid_height,
+        discretize=100,
+    ):
+        super().__init__(name, geometry, flux_time, gas, liquid, discretize)
+
+        # Define liquid height
+        self.liquid_height = Function(
+            liquid_height, "Time (s)", "volume (m³)", "linear"
```

Comment:
Is it volume or height?

Reply:
Fixed on 80daa46310121108ea6557f44e5c4663c2892ef3.